### PR TITLE
feat: refetch wallet balances on interval

### PR DESCRIPTION
**Pull request by 0xpatrickdev**
- feat: refetch wallet balances on interval (30s)
- test: add e2e tests for all api queries 
- ci: use [agoric-3-proposals](https://github.com/Agoric/agoric-3-proposals/) docker image for e2e tests
- see https://github.com/0xpatrickdev/agoric-gov-proposal-hub/issues/41

### CODE REVIEW

**Review by dckc — APPROVED** [review]
e2e tests look nifty!

I haven't looked at the feature closely nor tested it.

File: .github/workflows/pr.yml [hunk 1]
```diff
@@ -21,5 +21,46 @@ jobs:
...
+            if [ "$code" == "12" ] && [ "$message" == "Not Implemented" ]; then
+              echo "Service is ready!";
+              break;
+            fi;
+            sleep 10;
+            timeout=$((timeout-10));
+            if [ "$timeout" -le 0 ]; then
+              echo "Service did not become ready in time";
+              exit 1;
+            fi;
```
[the remark this answers is not in the document]
Comment (0xpatrickdev) [reply .github/workflows/pr.yml]: This was something chatgpt helped me write, thank you for raising as I now see the issue (2 timers).

How does 7389fe4 look?

File: .github/workflows/pr.yml [hunk 2]
```diff
@@ -21,5 +21,46 @@ jobs:
...
+      - name: Wait for Docker container to be ready
+        run: |
+          timeout 300 bash -c '
+          echo "Waiting for the agoric service to be ready...";
+          while true; do
+            response=$(curl --silent http://localhost:1317);
+            echo "Checking service response: $response";
+            code=$(echo $response | jq .code);
+            message=$(echo $response | jq -r .message);
+            if [ "$code" == "12" ] && [ "$message" == "Not Implemented" ]; then
```
[the remark this answers is not in the document]
Comment (0xpatrickdev) [reply .github/workflows/pr.yml]: `:1317/status` is not implemented and `:26657/status` always seems to return block 11 🤔 . 

Updated in https://github.com/DCFoundation/cosmos-proposal-builder/commit/7389fe48da5df9d76eaacd21a958af14e5ef7617 to check `block_height` from `/abci_query`. Height 1001 is when all of the proposals seem to be done.

File: README.md [hunk 3]
```diff
@@ -27,6 +27,46 @@ yarn build
...
+yarn test:unit  # Run all tests
+```
+
+### End-to-End Tests
+
+End-to-end tests simulate an actual blockchain. Make sure you have [Docker](https://docs.docker.com/engine/install/) installed to run these tests.
+
+#### 1. Start the Blockchain Node
+
+We use a fork of [agoric-3](https://github.com/Agoric/agoric-sdk/releases) for our tests. To start the simulation, run the following command which pulls and runs the [`agoric-3-proposals`](https://github.com/Agoric/agoric-3-proposals/) image, setting up an environment similar to the Agoric mainnet:
```
[the remark this answers is not in the document]
Comment (0xpatrickdev) [reply README.md:48]: Agree this is a bit awkward. Open to a better copy/link suggestion if you have one!

File: src/lib/queries.e2e.tsx [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        "fee_unit_price",
+        "bootstrap_vat_config",
+        "power_flag_fees",
+        "queue_max",
+      ]
+    `);
+      expect(result.current.data).toMatchSnapshot();
+    });
+  });
+  describe("accountBalancesQuery Query", () => {
```
[the remark this answers is not in the document]
Comment (0xpatrickdev) [reply src/lib/queries.e2e.tsx]: [98bad69](https://github.com/DCFoundation/cosmos-proposal-builder/pull/7/commits/98bad693d146793569f6468d4e2a78182df5525a)

File: .github/workflows/pr.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (dckc) [reply .github/workflows/pr.yml]: I have avoided grokking shell stuff such as `if ! timeout ...`.
 - C heritage would suggest that means "if timeout exits 0", but
 - the code here seems to treat it as a problem, i.e. as if it means "if timeout exits non-zero".

The latter interpretation seems to have 126 endorsements: https://stackoverflow.com/a/26475880/7963
I remain confused. Does `if X` mean `if X is 0`??? I guess so... or rather: `if X exits 0`.

File: README.md (hunk 3 above)
Comment (dckc) [inline README.md:48]: I'm not entirely clear whether you mean to refer to the `agoric-3` chain or to some release or releases. Assuming you mean the chain...

```suggestion
We use a simulation of `agoric-3` for our tests. To start the simulation, run the following command which pulls and runs the [`agoric-3-proposals`](https://github.com/Agoric/agoric-3-proposals/) image, setting up an environment similar to the Agoric mainnet:
```

[the remark this answers is not in the document]
Comment (0xpatrickdev) [reply README.md]: Fixed `[c448cc0](https://github.com/DCFoundation/cosmos-proposal-builder/pull/7/commits/c448cc0b017bb629f7738c474930949133790347)`

File: .github/workflows/pr.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (0xpatrickdev) [reply .github/workflows/pr.yml]: > Does if X mean if X is 0??? I guess so... or rather: if X exits 0.

Yes, the convention means `if command X exits with a status of 0`, read as `if X succeeds`.

I removed it and have something working as of [6921fb2](https://github.com/DCFoundation/cosmos-proposal-builder/pull/7/commits/6921fb209ff767b14b8f6a53d5a9a17c52c23659).

[CI Output](https://github.com/DCFoundation/cosmos-proposal-builder/actions/runs/7416332252/job/20181121612?pr=7#step:7:25): 

```bash
Waiting for the Agoric service to be fully ready...
Target block height: 1001
Waiting for last block height to reach 1001. Current height: 980
Waiting for last block height to reach 1001. Current height: 991
Service is ready! Last block height: 1001
```

Inclined to mark this resolved and merge, but further feedback/simplification is welcome!

File: .github/workflows/pr.yml [hunk 5]
```diff
@@ -21,5 +21,45 @@ jobs:
...
         run: yarn ts:check
-      - name: yarn test
-        run: yarn test
+      - name: yarn test:unit
+        run: yarn test:unit
+  e2e:
+    runs-on: ubuntu-latest
+    services:
+      agoric:
+        image: ghcr.io/agoric/agoric-3-proposals:main # fork of agoric-3
```
Comment (0xpatrickdev) [inline .github/workflows/pr.yml:30]:
```suggestion
        image: ghcr.io/agoric/agoric-3-proposals:main # simulation of agoric-3
```

File: .github/workflows/pr.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (dckc) [reply .github/workflows/pr.yml]: go for it

**Pull request merged**